#955 clear out uses of addAction in tests

bradybray opened the pull request:
@PhilMiller @lifflander 
There should be some discussion about the test case test_pending_send.cc. That is a special case and was not amenable to the typical case of removing add action and placing runInEpochCollective. Jonathan and I had a call on Sunday to discuss the implementation in this request. The issue is that the runInEpochCollective has no support for actions that are emplaced for execution later. Instead it just runs a unit of work and moves on. That does not work in the test_pending_send case as it is a sort of prepare and wait sort of send instead of send at the immediate moment.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

PhilMiller commented on tests/unit/active/test_active_send.cc:
```diff
@@ -132,18 +132,20 @@ TEST_F(TestActiveSend, test_type_safe_active_fn_send) {
+        auto msg = makeMessage<TestMsg>();
+        theMsg()->sendMsg<TestMsg, test_handler>(1, msg.get());
+      }
+	}
```
Looks like a slightly off indentation here

PhilMiller commented on tests/unit/active/test_pending_send.cc:
```diff
@@ -85,18 +85,18 @@ TEST_F(TestPendingSend, test_pending_send_hold) {
+  // theTerm->addAction([&done] { done = true; });
   theTerm()->finishedEpoch(ep);
 
   // It should not break out of this loop because of `done`, thus `k` is used to
```
This comment needs to be updated to match the changed logic

PhilMiller commented on tests/unit/active/test_pending_send.cc:
```diff
@@ -85,18 +85,18 @@ TEST_F(TestPendingSend, test_pending_send_hold) {
 
   theMsg()->popEpoch(ep);
-  theTerm()->addAction(ep, [&done] { done = true; });
+  // theTerm->addAction([&done] { done = true; });
```
Please delete this rather than just leaving it commented out

PhilMiller commented on tests/unit/active/test_pending_send.cc:
```diff
@@ -105,15 +105,16 @@ TEST_F(TestPendingSend, test_pending_send_hold) {
   pending.clear();
 
-  do vt::runScheduler(); while (not done);
+  // do vt::runScheduler(); while (not done);
```
Ditto commented code

PhilMiller commented on tests/unit/collection/test_checkpoint.extended.cc:
```diff
@@ -128,7 +128,7 @@ static void runInEpoch(std::function<void()> fn) {
   vt::theTerm()->finishedEpoch(ep);
   bool done = false;
-  vt::theTerm()->addAction(ep, [&]{ done = true; });
+  vt::runInEpochCollective([&]{ done = true; });
   vt::theSched()->runSchedulerWhile([&] { return not done; });
```
This seems to be a mistake. The entirety of the function `runInEpoch` could be replaced at the call site(s) with `runInEpochCollective`

PhilMiller commented on tests/unit/collection/test_insert.extended.cc:
```diff
@@ -89,126 +89,132 @@ static constexpr int32_t const num_elms_per_node = 8;
     /// ::fmt::print("num inserted={}\n", num_inserted);
     // Relies on default mapping equally distributing
     EXPECT_EQ(num_inserted, num_elms_per_node);
     num_inserted = 0;
```
Looks like this didn't get un-indented with the block nesting change

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -54,15 +54,14 @@ using TestMPICollective = TestParallelHarness;
   });
+  EXPECT_TRUE(done);
 
   while (not vt::rt->isTerminated()) {
```
This scheduler loop is no longer necessary

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -76,40 +75,41 @@ TEST_F(TestMPICollective, test_mpi_collective_2) {
+  EXPECT_EQ(bcast_val, 29);
+  EXPECT_EQ(reduce_val_out, num_nodes);
 
   while (not vt::rt->isTerminated()) {
```
Again, scheduler loop no longer necessary

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -193,54 +193,54 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
-    });
-  };
+  
+  vt::runInEpochCollective([&]{
```
Could this move to just wrap around the actual calls in the `if(is_even)`?

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -193,54 +193,54 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
 
   auto epoch = theTerm()->makeEpochCollective();
```
With `runInEpochCollective` in use, there's no need to manually create and activate an epoch; please delete it.

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -193,54 +193,54 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
+  EXPECT_EQ(reduce_val_out, num_nodes);
+  finished_spin = true;
 
   while (not finished_spin) {
```
The `finished_spin` variable and this scheduler loop should not be necessary. As the code has been modified, it's also a no-op, since the variable is set just before the loop.

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -193,54 +193,54 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
+  finished_spin = true;
 
   while (not finished_spin) {
     runScheduler();
```
Please put the broadcast below inside a call to `runInEpochCollective`, rather than having its execution rely on the scheduler being run to global termination during runtime finalization

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -193,54 +193,54 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
+        MPI_Barrier(comm);
+        run_order[done++] = 3;
+      });
+    };
```
@lifflander I think this test will only pass as modified by accident. I believe it actually illustrates a latent mistake we made in the async MPI collectives - they should produce on the active epoch when queued up, and consume on that same epoch when actually executed. The reduction to `CollectiveAlg::runCollective` will take care of things in the simplest cases, but I think it can go wrong when there are multiple concurrent collectives potentially in flight, with several of them queuing and returning.

PhilMiller replied:
Opened as #986.

PhilMiller commented:
For your commit messages, please put the issue number right at the front of the subject, without a "Task" prefix

PhilMiller commented:
Note to self: this PR touches all of the files assigned

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

bradybray commented:
@lifflander @JacobDomagala @cz4rs @PhilMiller Is something wrong with my pull request that I did not notice? My review is still open.

PhilMiller commented:
> @lifflander @JacobDomagala @cz4rs @PhilMiller Is something wrong with my pull request that I did not notice? My review is still open.

Nothing necessarily wrong. I was on vacation most of last week, and didn't get through everything in detail. Also, I don't want to jump to merging things before encouraging others working on the same topic to have a look.

So, no worried, it's not about your code or any problems in it, just getting enough attention mustered

cz4rs commented on tests/unit/collection/test_insert.extended.cc:
```diff
@@ -89,126 +89,132 @@ static constexpr int32_t const num_elms_per_node = 8;
-    EXPECT_EQ(num_inserted, num_elms_per_node);
-    num_inserted = 0;
   });
+  /// ::fmt::print("num inserted={}\n", num_inserted);
```
minor: most of the commented out `print`s can probably be removed, here `EXPECT_EQ` will cover the printing if something is wrong

cz4rs commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -191,60 +183,48 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
-      vt_print(barrier, "run MPI_barrier\n");
-      MPI_Barrier(comm);
-      run_order[done++] = 3;
+	  vt_print(barrier, "run MPI_barrier\n");
```
formatting: I think you have a tab mixed in with spaces here?

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -191,60 +183,48 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
 
   auto op2 = [&]{
     scope2.mpiCollectiveAsync([&done,&reduce_val_out]{
-      auto comm = theContext()->getComm();
+    auto comm = theContext()->getComm();
       int val_in = 1;
-      vt_print(barrier, "run MPI_Allreduce\n");
-      MPI_Allreduce(&val_in, &reduce_val_out, 1, MPI_INT, MPI_SUM, comm);
-      run_order[done++] = 2;
+    vt_print(barrier, "run MPI_Allreduce\n");
+    MPI_Allreduce(&val_in, &reduce_val_out, 1, MPI_INT, MPI_SUM, comm);
+    run_order[done++] = 2;
     });
   };
 
   auto op3 = [&]{
     scope3.mpiCollectiveAsync([&done]{
       auto comm = theContext()->getComm();
-      vt_print(barrier, "run MPI_barrier\n");
-      MPI_Barrier(comm);
-      run_order[done++] = 3;
+	  vt_print(barrier, "run MPI_barrier\n");
+	  MPI_Barrier(comm);
+	  run_order[done++] = 3;
```
Indentation has changed here since my last look, but it now seems to off in a different way.

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -191,60 +183,47 @@ TEST_F(TestMPICollective, test_mpi_collective_4) {
-      run_order[done++] = 2;
+    vt_print(barrier, "run MPI_Allreduce\n");
+    MPI_Allreduce(&val_in, &reduce_val_out, 1, MPI_INT, MPI_SUM, comm);
+    run_order[done++] = 2;
```
As displayed on Github's PR interface, indentation still looks a little bit off here.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_insert.extended.cc  4
- tests/unit/collection/test_destroy.cc  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

PhilMiller commented on tests/unit/collection/test_checkpoint.extended.cc:
```diff
@@ -178,34 +167,34 @@ TEST_F(TestCheckpoint, test_checkpoint_1) {
-    for (int i = 0; i < 5; i++) {
-      runInEpoch([&]{
+    vt::runInEpochCollective([&]{
+      for (int i = 0; i < 5; i++) {
```
Why are we transposing the loop and the execution? I expect the test is fine either way, but it would seem to maybe be off-topic for this PR, unless I'm missing the reason for it.

bradybray replied:
@PhilMiller When I tried resolve it inside the loop it would cause a segfault.

lifflander replied:
I changed this back. It doesn't segfault for me if I move it there. Not sure what you were seeing.

lifflander commented:
@bradybray I rebased this on develop and fixed some indentation/tabbing issues. You might want to set your editor to not include tabs in the source code (spaces only). Also, I noticed trailing whitespace that I removed.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)

bradybray commented:
> @bradybray I rebased this on develop and fixed some indentation/tabbing issues. You might want to set your editor to not include tabs in the source code (spaces only). Also, I noticed trailing whitespace that I removed.

Thanks for the fix. I recently changed notepad++ to use the projects spacing conventions. I'll also make sure to watch out for the white space issue.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_destroy.cc  1
- tests/unit/collection/test_insert.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5986060&bid=19629228)
